src/binary/Builtin.cc

```diff
@@ -222,16 +222,37 @@ extern "C" {
   make_crypt_salt (const char* crypt_prefix, int crypt_rounds)
   {
     // use gensalt own auto entropy
-    const char *entropy = NULL;
-    const size_t entropy_len = 0;
+    char* retval = crypt_gensalt_ra (crypt_prefix, crypt_rounds, NULL, 0);
 
-    char* retval = crypt_gensalt_ra (crypt_prefix, crypt_rounds, entropy, entropy_len);
-
-    if (!retval)
+    // auto entropy might not be supported in some older systems (15.2 and older),
+    // if we get the "Invalid argument" error then read some entropy from
+    // /dev/urandom and try again
+    if (!retval && errno == EINVAL)
     {
-      y2error ("Unable to generate a salt, check your crypt settings.\n");
+      const char* device = "/dev/urandom";
+
+      int fd = open (device, O_RDONLY);
+      if (fd < 0)
+      {
+        y2error ("Can't open %s for reading: %s\n", device, strerror (errno));
+        return 0;
+      }
+
+      char entropy[16];
+      const size_t entropy_len = sizeof(entropy);
+      if (read_loop (fd, entropy, entropy_len) != entropy_len)
+      {
+        close (fd);
```

Comment:
why close fd only if read failed? does not it leak fd?

Reply:
Ah, good point... I just copied the original code...